File: internal/models/user_permission.go
package models

import (
	"github.com/google/uuid"
	"github.com/juevigrace/diva-server/internal/models/responses"
	"github.com/juevigrace/diva-server/storage"
)

type UserPermission struct {
	Permission Permission
	UserID     uuid.UUID
	GrantedBy  *uuid.UUID
	Granted    bool
	GrantedAt  int64
	// TODO: change expiration time for enum with fixed times
	ExpiresAt *int64
	UpdatedAt int64
}

func (up *UserPermission) Response() *responses.UserPermissionResponse {
	var grantedBy *string
	if up.GrantedBy != nil {
		grantedBy = new(string)
		*grantedBy = up.GrantedBy.String()
	}

	return &responses.UserPermissionResponse{
		PermissionID: up.Permission.ID.String(),
		GrantedBy:    grantedBy,
		Granted:      up.Granted,
		GrantedAt:    up.GrantedAt,
		ExpiresAt:    up.ExpiresAt,
		UpdatedAt:    up.UpdatedAt,
	}
}

func (up *UserPermission) DBCreate(userID uuid.UUID) *storage.CreateUserPermissionParams {
	return &storage.CreateUserPermissionParams{
		PermissionID: up.Permission.ID,
		UserID:       userID,
		GrantedBy:    up.GrantedBy,
		Granted:      up.Granted,
		ExpiresAt:    up.ExpiresAt,
	}
}

func (up *UserPermission) DBUpdate(userID uuid.UUID) *storage.UpdateUserPermissionParams {
	return &storage.UpdateUserPermissionParams{
		PermissionID: up.Permission.ID,
		UserID:       userID,
		Granted:      up.Granted,
		ExpiresAt:    up.ExpiresAt,
	}
}

func UserPermissionFromDB(row *storage.DivaUserPermission, perm *Permission) *UserPermission {
	return &UserPermission{
		Permission: *perm,
		UserID:     row.UserID,
		GrantedBy:  row.GrantedBy,
		Granted:    row.Granted,
		GrantedAt:  row.GrantedAt,
		ExpiresAt:  row.ExpiresAt,
		UpdatedAt:  row.UpdatedAt,
	}
}

File: client/packages/shared/app-lib/diva-lib-models-core/src/commonMain/kotlin/io/github/juevigrace/diva/lib/models/user/permissions/UserPermission.kt
@file:OptIn(ExperimentalJsExport::class)
@file:DivaJsExport

package io.github.juevigrace.diva.lib.models.user.permissions

import io.github.juevigrace.diva.core.DivaJsExport
import io.github.juevigrace.diva.core.None
import io.github.juevigrace.diva.core.Option
import io.github.juevigrace.diva.lib.models.api.user.permissions.UserPermissionResponse
import io.github.juevigrace.diva.lib.models.permission.Permission
import io.github.juevigrace.diva.lib.models.permission.PermissionAction
import io.github.juevigrace.diva.lib.models.user.Role
import kotlin.js.ExperimentalJsExport

data class UserPermission(
    val permission: Permission,
    val grantedBy: Option<String> = None,
    val granted: Boolean,
    val grantedAt: Option<Long> = None,
    val expiresAt: Option<Long> = None,
    val updatedAt: Long,
) {
    companion object {
        fun fromResponse(response: UserPermissionResponse): UserPermission {
            return UserPermission(
                permission = Permission(
                    id = response.permissionId,
                    name = "",
                    description = "",
                    action = PermissionAction.PERMISSION_NONE,
                    roleLevel = Role.USER,
                    createdAt = response.grantedAt ?: response.updatedAt,
                    updatedAt = response.updatedAt,
                ),
                grantedBy = Option.of(response.grantedBy),
                granted = response.granted,
                grantedAt = Option.of(response.grantedAt),
                expiresAt = Option.of(response.expiresAt),
                updatedAt = response.updatedAt,
            )
        }
    }
}
